org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaRunner.java

```diff
@@ -92,6 +111,26 @@ public void compileAndRunUserProgram(UserProgram userProgram, Principal principa
 
     // ensure System.out is reset
     System.setOut(System.out);
+    System.setIn(System.in);
+  }
+
+  /**
+   * Write the given userInput to the user's (principal's) input stream
+   *
+   * @param userInput the given input directed to System.in
+   * @param principal the user's identification
+   */
+  public void passInputToRuntime(UserInput userInput, Principal principal) {
+    // System.in expects input to end with a lineSeparator. We add that to allow the user program to
+    // continue.
+    byte[] input = (userInput.getInput() + System.lineSeparator()).getBytes(StandardCharsets.UTF_8);
```

Comment:
For my education, is the plan for the client to send a UserInput message when a newline is encountered (and the newline will be stripped before sending the message)?

Reply:
Hmm yeah, the way this is written is a little odd. Initially, I just added this in because the program kept hanging when I forgot to hit "enter" before hitting the button to send the message. To answer your question: we haven't planned out the client message format yet. A couple things that make me think we should keep some version of this line:
- We can't be sure that all clients platform & browser combinations will pass a line separator that this Linux server will recognize
- It's likely we'll find some creative way of passing input from the client in the future where it wouldn't logically make sense to add a line separator (i.e. from a form or similar). It would be difficult to always remember to add a line separator in those cases.

Perhaps a better way to write this would be to check for a system-recognized line separator and add one if none exists.

Reply:
That sounds reasonable.  I think part of my question was whether the protocol for sending the input stream is line-based (i.e. input is sent line by line).  I think this simplifies things, but does impose some constraints (e.g. can't submit input without hitting enter).  If the protocol is line-based, we should be clear about whether the newline should or should not be sent by the client (or perhaps that we want to be lenient and accept both but I would not recommend that, at least to start).